tasmota/tasmota_xdrv_driver/xdrv_16_tuyamcu_upgrade.ino

```diff
@@ -0,0 +1,390 @@
+/*
+  xdrv_16_tuyamcu_upgrade.ino - Tuya MCU upgrade support for Tasmota
+
+  Copyright (C) 2023
+
+  This program is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifdef USE_LIGHT
+#ifdef USE_TUYA_MCU
+#ifdef USE_TUYA_MCU_UPGRADE
+
+#include <include/xdrv_16_tuyamcu_upgrade.h>
+
+#include <t_bearssl_hash.h>
+#include <memory>
+
+#ifdef ESP8266
+extern "C" uint32_t _FS_start;     // ... depending on core-sdk-version
+#elif defined ESP32
+#ifdef USE_UFILESYS
+extern FS *ufsp;
+extern char *fileOnly (char *);
+#else
+#error USE_TUYA_MCU_UPGRADE will only compile with active UFS (define USE_UFILESYS)
+#endif
+#endif
+
+TuyaUpgBuffer::TuyaUpgBuffer (void) {
+
+}
+
+TuyaUpgBuffer::~TuyaUpgBuffer (void) {
+#ifdef ESP32
+  if (_fname) {
+    delete[] _fname;
+    _fname = nullptr;
+  }
+#endif
+  if (_buffer) {
+    delete[] _buffer;
+    _buffer = nullptr;
+  }
+  if (_md5_context) {
+    delete _md5_context;
+    _md5_context = nullptr;
+  }
+  if (_current_packet) {
+    delete[] _current_packet;
+    _current_packet = nullptr;
+  }
+}
+
+bool TuyaUpgBuffer::init (uint32_t len, char* checksum, char* filename) {
+  reset();
+#ifdef ESP8266
+  // prepare to use OTA-partition
+  _start = (ESP.getSketchSize() + _flash_sector_size - 1) & (~(_flash_sector_size - 1));
+  _end = (uint32_t)&_FS_start - 0x40200000;
```

Comment:
Dangerous as it can overwrite ESP8266 settings located in eight banks below _FS_start. See Tasmota OTA function which makes sure the latests settings are stored safely before the Tasmota OTA is allowed to start,

Reply:
Fixed this by commit [c2e284b](https://github.com/arendst/Tasmota/pull/19877/commits/c2e284b565a9f56fd39f8b90a5f4d1bcb53415f8)

Reply:
Maybe we should take a look at FLOG::init(void) in [support_flash_log.ino](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_support/support_flash_log.ino). That was my original source of code snippet copy to prepare the OTA part.

Reply:
Yeah, this code part is rather old. At this time many API / functions in Tasmota just did not exists.